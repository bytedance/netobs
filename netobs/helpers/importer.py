# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import importlib.util
import sys
import uuid
from importlib import import_module
from typing import Any


def import_module_or_file(module_name: str) -> Any:
    """Import a python module or a python file.

    Args:
        module_name: the name of the module or file.
            If it ends with ".py", it will be considered as a file, otherwise module.

    Returns:
        Contents of the module.

    Raises:
        OSError: Python ifle not found.
    """
    if module_name.endswith(".py"):
        # generate unique module name
        module_id = "netobs_" + str(uuid.uuid4()).replace("-", "_")
        # `imp` is deprecated. Using `importlib` way
        spec = importlib.util.spec_from_file_location(module_id, module_name)
        if spec is None or spec.loader is None:
            raise OSError(f"Failed to load {module_name}")
        module = importlib.util.module_from_spec(spec)
        sys.modules[module_id] = module
        spec.loader.exec_module(module)
    else:
        module = import_module(module_name)
    return module
