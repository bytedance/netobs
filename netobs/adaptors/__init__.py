# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""This file holds common types and interfaces for networks.

All networks need to implement or integrate with NetworkAdaptor class to work with this
codebase. For example,

```python
@register_pytree_node_class
class AwesomeNetAdaptor(NetworkAdaptor):
    def call_network(...):
        ...

    # And other methods ...
```

Usually, to restore and evaluate the network, one may need more arguments than those
passed through arguments of the methods. Then `__init__` is the perfect place to
initialize your adaptor with some configurations and many more. There is no restriction
to the `__init__` method, since all the functions only take an instance of adaptor.

And an instance of `AwesomeNetAdaptor` should then be passed to `evaluate_observable`.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any, Callable, Generic, Protocol, TypeVar

import jax
from jax import numpy as jnp
from typing_extensions import Self

from netobs.helpers.grad import grad_with_system
from netobs.systems import System

X = TypeVar("X", bound=System)
D = TypeVar("D")


class WalkingStep(Protocol[D]):
    """MCMC walking step function."""

    def __call__(
        self,
        key: jnp.ndarray,
        params: jnp.ndarray,
        electrons: jnp.ndarray,
        aux_data: D,
    ) -> tuple[jnp.ndarray, D]:
        pass


class NetworkAdaptor(ABC, Generic[X]):
    """Common interface for different networks.

    Args:
        config: Anything telling the adaptor how to load network config.
            But a string is preferred because it makes life easier for command line.
        args: Command line args to modify config.
    """

    def __init__(self, config: Any, args: list[str]) -> None:
        super().__init__()

    def tree_flatten(self) -> tuple[tuple, Self]:
        """Flatten the class as PyTree.

        Since the adaptor is usually static and doesn't store any data, we just return
        empty `children` list and `self` as `aux_data`.

        Useful when JIT-ting and `pmap`-ping some methods of the adaptor class.

        Returns:
            A tuple of children (which itself is an empty tuple) and the adaptor itself.
        """
        return ((), self)

    @classmethod
    def tree_unflatten(cls, aux_data: Self, children: tuple) -> Self:
        """Unflatten the class as PyTree.

        Args:
            aux_data: the adaptor itself
            children: ignored

        Returns:
            The adaptor it self.
        """
        del children
        return aux_data

    @abstractmethod
    def restore(self, restore_option: Any) -> tuple[Any, jnp.ndarray, X, Any]:
        """Restore checkpoint with given options.

        Args:
            restore_option: anything needed for restoring the network from checkpoint.
                It's just the `options.network_restore` passed to `evaluate_observable`.

        Returns:
            Restored data:
            - params
            - system
            - aux_data
        """

    @abstractmethod
    def call_signed_network(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: X
    ) -> tuple[jnp.ndarray, jnp.ndarray]:
        """Get the sign and log magnitude of the wavefunction.

        If the network is complex-valued, return log phase and log psi.

        Args:
            params: network parameters.
            electrons: electron positions, shape (nelectrons*ndim), where ndim is the
                dimensionality of the system.
            system: system containing atomic info.

        Returns:
            The sign (or phase) and log magnitude of the wavefunction.
        """

    def call_network(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: X
    ) -> jnp.ndarray:
        """Get the log magnitude of the wavefunction.

        Neither sign nor phase factor should be included.

        Args:
            params: network parameters.
            electrons: electron positions, shape (nelectrons*ndim), where ndim is the
                dimensionality of the system.
            system: system containing atomic info.

        Returns:
            The log magnitude of the wavefunction.
        """
        return self.call_signed_network(params, electrons, system)[1]

    @abstractmethod
    def make_walking_step(
        self,
        batch_log_psi: Callable[[jnp.ndarray, jnp.ndarray, X], jnp.ndarray],
        steps: int,
        system: X,
    ) -> WalkingStep:
        """Make MCMC walking step.

        Args:
            batch_log_psi: Batched function returning log psi.
                (params, (nbatch, nelec*ndim), system) -> (nbatch,)
                We receive it as input because the importance sampling can be based on
                another psi_G instead of base psi_T.
            steps: Number of MCMC steps to run.
            system: system containing atomic info.

        Returns:
            Function with the same signature as WalkingStep
        """

    def make_burnin_step(
        self,
        batch_log_psi: Callable[[jnp.ndarray, jnp.ndarray, X], jnp.ndarray],
        steps: int,
        system: X,
    ) -> WalkingStep:
        """Make MCMC burnin step.

        By default, this is the same as make_walking_step.

        Args:
            batch_log_psi: Batched function returning log psi.
            steps: Number of MCMC steps to run.
            system: system containing atomic info.

        Returns:
            Function with the same signature as WalkingStep
        """
        return self.make_walking_step(batch_log_psi, steps, system)

    def call_local_energy(
        self, params: jnp.ndarray, key: jnp.ndarray, electrons: jnp.ndarray, system: X
    ) -> jnp.ndarray:
        """Get local energy.

        Args:
            params: Network params.
            key: JAX random keys.
            electrons: Electon positions.
            system: System containing atomic info.

        Returns:
            Local energy at this point.
        """
        key0, key1 = jax.random.split(key)
        ke = self.call_local_kinetic_energy(params, key0, electrons, system)
        pe = self.call_local_potential_energy(params, key1, electrons, system)
        return ke + pe

    @abstractmethod
    def call_local_kinetic_energy(
        self, params: jnp.ndarray, key: jnp.ndarray, electrons: jnp.ndarray, system: X
    ) -> jnp.ndarray:
        """Get local kinetic energy.

        Args:
            params: Network params.
            key: JAX random keys.
            electrons: Electon positions.
            system: System containing atomic info.

        Returns:
            Local kinetic energy at this point.
        """

    @abstractmethod
    def call_local_potential_energy(
        self, params: jnp.ndarray, key: jnp.ndarray, electrons: jnp.ndarray, system: X
    ) -> jnp.ndarray:
        """Get local potential energy.

        Args:
            params: Network params.
            key: JAX random keys.
            electrons: Electon positions.
            system: System containing atomic info.

        Returns:
            Local potential energy at this point.
        """

    def make_network_grad(self, arg: str, jaxfun: Callable = jax.grad):
        """Create gradient function of network.

        Useful when the gradient handling is different in your network.
        Note that if your network is complex-valued, you don't need to implement
        this function since `self.call_network` must be real.

        Args:
            arg: the name of arguments to calculate gradient, e.g. "electrons"/"atoms".
            jaxfun: the type of gradient to take, e.g. `jax.grad`.

        Returns:
            The gradient function.
        """
        return grad_with_system(self.call_network, arg, jaxfun=jaxfun)  # type: ignore

    def make_signed_network_grad(self, arg: str, jaxfun: Callable = jax.grad):
        """Create gradient function of network, taking sign into consideration.

        Useful when the gradient handling is different in your network,
        e.g. the network is complex-valued.

        Args:
            arg: the name of arguments to calculate gradient, e.g. "electrons"/"atoms".
            jaxfun: the type of gradient to take, e.g. `jax.grad`.

        Returns:
            The gradient function.
        """
        return grad_with_system(self.call_network, arg, jaxfun=jaxfun)  # type: ignore

    def make_local_energy_grad(self, arg: str, jaxfun: Callable = jax.grad):
        """Create gradient function of local energy.

        Useful when the gradient handling is different in your network,
        e.g. the network is complex-valued.

        Args:
            arg: the name of arguments to calculate gradient, e.g. "electrons"/"atoms".
            jaxfun: the type of gradient to take, e.g. `jax.grad`.

        Returns:
            The gradient function.
        """
        return grad_with_system(self.call_local_energy, arg, jaxfun=jaxfun)  # type: ignore


# Utility protocols
class LogPsiNet(Protocol):
    def __call__(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: System
    ) -> jnp.ndarray:
        pass
