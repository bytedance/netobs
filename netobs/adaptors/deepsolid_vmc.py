# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from functools import partial
from pathlib import Path
from typing import Any, Callable

import DeepSolid.hamiltonian
import DeepSolid.hf
import DeepSolid.network
import DeepSolid.qmc
import DeepSolid.supercell
import jax
from DeepSolid.ewaldsum import EwaldSum
from DeepSolid.supercell import get_supercell_copies
from jax import numpy as jnp
from jax.tree_util import register_pytree_node_class
from typing_extensions import TypedDict

from netobs.adaptors import NetworkAdaptor, WalkingStep
from netobs.helpers.asbl_config import absl_config
from netobs.helpers.grad import grad_with_system, make_kinetic_energy
from netobs.systems.solid import MinimalImageDistance, SolidSystem


class DeepSolidVMCAuxData(TypedDict):
    """Auxiliary data for DeepSolid."""

    mcmc_width: jnp.ndarray


@register_pytree_node_class
class DeepSolidVMCAdaptor(NetworkAdaptor[SolidSystem]):
    def __init__(self, config: Any, args: list[str]) -> None:
        super().__init__(config, args)
        self.config = absl_config(config, args) if isinstance(config, str) else config

    def restore(
        self, ckpt_file: str | None = None, check_shape: bool = True
    ) -> tuple[Any, jnp.ndarray, SolidSystem, DeepSolidVMCAuxData]:
        cfg = self.config
        restore_path = Path(cfg.log.restore_path)
        ckpt_filename = (
            str(restore_path / f"qmcjax_ckpt_{cfg.optim.iterations-1:06d}.npz")
            if ckpt_file is None
            else ckpt_file
        )
        with open(ckpt_filename, "rb") as f:
            ckpt_data = jnp.load(f, allow_pickle=True)
            data = ckpt_data["data"]
            params = ckpt_data["params"].tolist()
            mcmc_width = jnp.array(ckpt_data["mcmc_width"].tolist())

        cards = jax.local_device_count()
        if check_shape and cards != data.shape[0]:
            print(
                f"Converting checkpoint with {data.shape[0]} devices to {cards} devices"
            )
            data = data.reshape(cards, -1, data.shape[-1])
            params = jax.tree_map(lambda x: jnp.repeat(x[:1], cards, axis=0), params)
            mcmc_width = jnp.repeat(mcmc_width[:1], cards, axis=0)

        simulation_cell = cfg.system.pyscf_cell

        hartree_fock = DeepSolid.hf.SCF(
            cell=simulation_cell,
            twist=jnp.array(cfg.network.twist),
        )
        hartree_fock.init_scf()
        klist = hartree_fock.klist

        self.atoms = atoms = simulation_cell.original_cell.atom_coords()
        charges = simulation_cell.original_cell.atom_charges()
        latvec = jnp.asarray(simulation_cell.original_cell.lattice_vectors())
        self.orig_dist = MinimalImageDistance(latvec)
        self.cell_size = simulation_cell.scale

        system = SolidSystem(
            atoms=atoms,
            charges=charges,
            spins=simulation_cell.nelec,
            ndim=cfg.system.ndim,
            latvec=simulation_cell.lattice_vectors(),
        )

        self.net_func = lambda params, electrons, atoms: DeepSolid.network.eval_func(
            params,
            electrons,
            atoms=atoms,
            klist=klist,
            simulation_cell=simulation_cell,
            spins=simulation_cell.nelec,
            envelope_type=cfg.network.detnet.envelope_type,
            full_det=cfg.network.detnet.full_det,
            distance_type=cfg.network.detnet.distance_type,
            method_name="eval_phase_and_slogdet",
        )

        self.local_kinetic = make_kinetic_energy(
            self.call_complex_network,
            DeepSolid.hamiltonian.local_kinetic_energy_real_imag,
        )
        self.local_potential = local_ewald_energy(simulation_cell)
        twist = jnp.mod(jnp.array(cfg.network.twist), 1.0)
        ks = jnp.dot(jnp.linalg.inv(simulation_cell.a), twist) * 2 * jnp.pi
        if simulation_cell._ecp:
            self.non_local_potential = DeepSolid.hamiltonian.non_local_energy(
                partial(self.call_signed_network, system=system),
                simulation_cell,
                minimal_imag_dis=MinimalImageDistance(simulation_cell.a),
                ecp_quadrature_id=cfg.system.ecp_quadrature_id,
                partition_number=sum(simulation_cell.nelec)
                if cfg.optim.ecp_partition_number == -1
                else cfg.optim.ecp_partition_number,
                ks=ks,
            )
        else:
            self.non_local_potential = lambda *_: 0

        return (
            params,
            data,
            system,
            DeepSolidVMCAuxData(mcmc_width=mcmc_width),
        )

    def call_local_kinetic_energy(
        self,
        params: jnp.ndarray,
        key: jnp.ndarray,
        electrons: jnp.ndarray,
        system: SolidSystem,
    ) -> jnp.ndarray:
        del key
        return sum(self.local_kinetic(params, electrons, system)) / self.cell_size

    def call_local_potential_energy(
        self,
        params: jnp.ndarray,
        key: jnp.ndarray,
        electrons: jnp.ndarray,
        system: SolidSystem,
    ) -> jnp.ndarray:
        del key
        ep = self.local_potential(electrons, system)
        # HACK: grad potential requires real-valued potential
        ecp = self.non_local_potential(params, electrons, system["atoms"]).real
        return (ep + ecp) / self.cell_size

    def call_signed_network(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: SolidSystem
    ) -> tuple[jnp.ndarray, jnp.ndarray]:
        return self.net_func(params, electrons, system["atoms"])

    def call_network(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: SolidSystem
    ) -> jnp.ndarray:
        _, output = self.net_func(params, electrons, system["atoms"])
        return jnp.real(output)

    def call_complex_network(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: SolidSystem
    ) -> jnp.ndarray:
        _, output = self.net_func(params, electrons, system["atoms"])
        return output

    def make_walking_step(
        self, batch_log_psi: Callable, steps: int, system: SolidSystem
    ) -> WalkingStep[DeepSolidVMCAuxData]:
        mcmc_step = DeepSolid.qmc.make_mcmc_step(
            lambda params, electrons: batch_log_psi(params, electrons, system),
            batch_per_device=1,  # useless
            latvec=system["latvec"],
            steps=steps,
        )

        def walk(
            key: jnp.ndarray,
            params: jnp.ndarray,
            electrons: jnp.ndarray,
            aux_data: DeepSolidVMCAuxData,
        ) -> tuple[jnp.ndarray, DeepSolidVMCAuxData]:
            new_data, _ = mcmc_step(params, electrons, key, aux_data["mcmc_width"])
            return new_data, aux_data

        return jax.pmap(walk)

    def make_signed_network_grad(self, arg: str, jaxfun: Callable = jax.grad):
        def complex_f(params, electrons, system):
            sign, slogdet = self.call_signed_network(params, electrons, system)
            return jnp.log(sign) + slogdet

        grad_f_real = grad_with_system(
            lambda *args: complex_f(*args).real, arg, args_before=1, jaxfun=jaxfun
        )
        grad_f_imag = grad_with_system(
            lambda *args: complex_f(*args).imag, arg, args_before=1, jaxfun=jaxfun
        )
        return lambda *args: grad_f_real(*args) + grad_f_imag(*args) * 1j

    def make_local_energy_grad(self, arg: str, jaxfun: Callable = jax.grad):
        grad_local_energy_real = grad_with_system(
            lambda *args: self.call_local_energy(*args).real,
            arg,
            args_before=2,
            jaxfun=jaxfun,
        )
        grad_local_energy_imag = grad_with_system(
            lambda *args: self.call_local_energy(*args).imag,
            arg,
            args_before=2,
            jaxfun=jaxfun,
        )
        return (
            lambda *args: grad_local_energy_real(*args)
            + grad_local_energy_imag(*args) * 1j
        )


def local_ewald_energy(simulation_cell):
    ewald = EwaldSum(simulation_cell)
    nelec = sum(simulation_cell.nelec)
    ewald.dist = MinimalImageDistance(ewald.latvec)
    Rpts = get_supercell_copies(
        simulation_cell.original_cell.lattice_vectors(), simulation_cell.S
    )

    def _local_ewald_energy(data: jnp.ndarray, system: SolidSystem):
        ewald.atom_coords = jnp.reshape(system["atoms"] + Rpts[:, None, :], (-1, 3))
        ii = ewald.ewald_ion() + ewald.ii_const
        ee, ei = ewald.ewald_electron(data)
        ee += ewald.ee_const(nelec)
        ei += ewald.ei_const(nelec)
        return jnp.asarray(ee) + jnp.asarray(ei) + jnp.asarray(ii)

    return _local_ewald_energy


DEFAULT = DeepSolidVMCAdaptor
