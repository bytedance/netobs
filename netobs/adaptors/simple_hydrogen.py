# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from functools import partial
from typing import Any, TypedDict

import jax
from jax import numpy as jnp
from jax.tree_util import register_pytree_node_class

from netobs.adaptors import NetworkAdaptor, WalkingStep
from netobs.systems.molecule import MolecularSystem


class HydrogenAuxData(TypedDict):
    mcmc_width: jnp.ndarray


@register_pytree_node_class
class SimpleHydrogen(NetworkAdaptor):
    """A simple adaptor for hydrogen atom wavefunction. Used for testing."""

    quality: float
    "How good the wavefunction is "

    def __init__(self, config: Any, args: list[str]) -> None:
        super().__init__(config, args)
        self.quality = float(config or 1)

    def restore(
        self, ckpt_file: str | None = None
    ) -> tuple[Any, jnp.ndarray, MolecularSystem, Any]:
        batch_size = int(ckpt_file or 2048)
        key = jax.random.PRNGKey(42)
        kr, kt, kp = jax.random.split(key, 3)
        shape = (1, batch_size, 1)
        r = jax.random.exponential(kr, shape)
        theta = jax.random.uniform(kt, shape, maxval=jnp.pi)
        phi = jax.random.uniform(kp, shape, maxval=2 * jnp.pi)
        data = jnp.concatenate(
            [
                r * jnp.sin(theta) * jnp.cos(phi),
                r * jnp.sin(theta) * jnp.sin(phi),
                r * jnp.cos(theta),
            ],
            axis=-1,
        )

        self.atoms = jnp.array([[0.0, 0.0, 0.0]])
        system = MolecularSystem(
            atoms=self.atoms, charges=jnp.array([1]), spins=(1, 0), ndim=3
        )
        aux_data = HydrogenAuxData(mcmc_width=jnp.array([0.6]))
        return (jnp.array([0]), data, system, aux_data)

    def call_signed_network(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: MolecularSystem
    ):
        del params
        # _, r_ae = calculate_r_ae(electrons, system)
        # r_ae = r_ae[0, 0, 0]  # 1st electron, 1st atom
        r_ae = jnp.linalg.norm(electrons - system["atoms"])
        return 1, -r_ae * self.quality - r_ae**2 * (1 - self.quality)

    def make_walking_step(
        self,
        steps: int,
        system: MolecularSystem,
    ) -> WalkingStep[HydrogenAuxData]:
        def walk(
            key: jnp.ndarray,
            params: jnp.ndarray,
            electrons: jnp.ndarray,
            aux_data: HydrogenAuxData,
        ) -> tuple[jnp.ndarray, HydrogenAuxData]:
            stddev = aux_data["mcmc_width"]

            def mh_update(_, data):
                x1, key, lp_1 = data
                key, subkey = jax.random.split(key)
                x2 = x1 + stddev * jax.random.normal(subkey, shape=x1.shape)  # proposal
                lp_2 = 2.0 * self.batch_network(params, x2, system)
                ratio = lp_2 - lp_1

                key, subkey = jax.random.split(key)
                cond = ratio > jnp.log(jax.random.uniform(key, shape=ratio.shape))
                x_new = jnp.where(cond[..., None], x2, x1)
                lp_new = jnp.where(cond, lp_2, lp_1)

                return x_new, key, lp_new

            logprob = 2.0 * self.batch_network(params, electrons, system)
            new_data, *_ = jax.lax.fori_loop(
                0, steps, mh_update, (electrons, key, logprob)
            )
            return new_data, aux_data

        return jax.pmap(walk, in_axes=0)

    def call_local_kinetic_energy(
        self,
        params: jnp.ndarray,
        key: jnp.ndarray,
        electrons: jnp.ndarray,
        system: MolecularSystem,
    ) -> jnp.ndarray:
        del key
        n = electrons.shape[0]
        eye = jnp.eye(n)
        grad_f = self.make_network_grad("electrons")

        def grad_f_closure(x):
            return grad_f(params, x, system)

        primal, dgrad_f = jax.linearize(grad_f_closure, electrons)

        return -0.5 * jax.lax.fori_loop(
            0, n, lambda i, val: val + dgrad_f(eye[i])[i], 0.0
        ) - 0.5 * jnp.sum(primal**2)

    def call_local_potential_energy(
        self,
        params: jnp.ndarray,
        key: jnp.ndarray,
        electrons: jnp.ndarray,
        system: MolecularSystem,
    ) -> jnp.ndarray:
        del params, key
        r_ae = jnp.linalg.norm(electrons - system["atoms"])
        # _, r_ae = calculate_r_ae(electrons, system)
        # r_ae = r_ae[0, 0, 0]  # 1st electron, 1st atom
        return -1 / r_ae

    @partial(jax.vmap, in_axes=(None, None, 0, None))
    def batch_network(self, params, electrons, system):
        return self.call_network(params, electrons, system)


DEFAULT = SimpleHydrogen
