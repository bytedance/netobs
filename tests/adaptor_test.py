# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from pathlib import Path

import jax
import pytest
from jax import numpy as jnp

from netobs.adaptors.simple_hydrogen import SimpleHydrogen
from netobs.helpers.importer import import_module_or_file

try:
    from netobs.adaptors.ferminet_vmc import FERMINET_VERSION

    HAS_FERMINET_MAIN = FERMINET_VERSION == "main"
except ImportError:
    HAS_FERMINET_MAIN = False


try:
    from netobs.adaptors.deepsolid_vmc import DeepSolidVMCAdaptor

    HAS_DEEPSOLID = True
except ImportError:
    HAS_DEEPSOLID = False


@pytest.mark.skipif(
    not jax.__version__.startswith("0.4."),
    reason="Different JAX version has different behaviors. Please use JAX 0.4.x",
)
class TestHydrogen:
    adaptor = SimpleHydrogen(0.999, [])

    def test_wf(self, snapshot):
        self.adaptor.quality = 1
        params, data, system, _ = self.adaptor.restore("1")
        data = data[0, 0]  # 1 data point, remove leading axis
        assert self.adaptor.call_network(params, data, system) == snapshot

    def test_potential(self, snapshot):
        params, data, system, _ = self.adaptor.restore("1")
        data = data[0, 0]
        assert (
            self.adaptor.call_local_potential_energy(params, None, data, system)
            == snapshot
        )

    def test_kinetic(self, snapshot):
        params, data, system, _ = self.adaptor.restore("1")
        data = data[0, 0]
        assert (
            self.adaptor.call_local_kinetic_energy(params, None, data, system)
            == snapshot
        )

    def test_walk(self, snapshot):
        params, data, system, aux_data = self.adaptor.restore("5")
        key = jax.random.PRNGKey(42)
        key = jax.random.split(key, 1)
        mcmc_step = self.adaptor.make_walking_step(1, system)
        assert mcmc_step(key, params, data, aux_data)[0] == snapshot


@pytest.mark.skipif(
    not HAS_FERMINET_MAIN,
    reason="FermiNet main is required to test it",
)
class TestFermiNet:
    @pytest.fixture(autouse=True)
    def _setup_adaptor(self):
        from ferminet.base_config import default
        from ferminet.utils import system

        from netobs.adaptors.ferminet_vmc import FermiNetMainAdaptor

        cfg = default()
        cfg.system.molecule = [
            system.Atom(symbol="H", coords=(0, 0, 0)),
        ]
        cfg.system.electrons = (1, 0)
        cfg.network.ferminet.hidden_dims = ((32, 4), (32, 4))
        self.adaptor = FermiNetMainAdaptor(cfg, [])
        self.net_retore = str(Path(__file__).parent / "data" / "H_atom.npz")
        params, data, system, aux_data = self.adaptor.restore(self.net_retore)
        self.params = jax.tree_util.tree_map(lambda x: x[0], params)
        self.data = data[0, 0]
        self.system = system
        self.aux_data = aux_data

    def test_wf(self, snapshot):
        logpsi = self.adaptor.call_network(self.params, self.data, self.system)
        assert logpsi == snapshot

    def test_potential(self, snapshot):
        potential = self.adaptor.call_local_potential_energy(
            self.params, None, self.data, self.system
        )
        assert potential == snapshot

    def test_kinetic(self, snapshot):
        kinetic = self.adaptor.call_local_kinetic_energy(
            self.params, None, self.data, self.system
        )
        assert kinetic == snapshot

    def test_walk(self, snapshot):
        # This one is pmapped
        params, data, system, aux_data = self.adaptor.restore(self.net_retore)
        key = jax.random.PRNGKey(42)
        key = jax.random.split(key, 1)
        mcmc_step = self.adaptor.make_walking_step(1, system)
        assert mcmc_step(key, params, data, aux_data)[0] == snapshot


@pytest.mark.skipif(
    not HAS_DEEPSOLID,
    reason="DeepSolid is required to test it",
)
class TestDeepSolid:
    @pytest.fixture(autouse=True)
    def _setup_adaptor(self):
        data_dir = Path(__file__).parent / "data"
        cfg = import_module_or_file(str(data_dir / "H_chain.py")).get_config()
        self.adaptor = DeepSolidVMCAdaptor(cfg, [])
        self.net_retore = str(data_dir / "H_chain.npz")
        params, data, system, aux_data = self.adaptor.restore(self.net_retore)
        self.params = jax.tree_util.tree_map(lambda x: x[0], params)
        self.data = data[0, 0]
        self.system = system
        self.aux_data = aux_data

    def test_wf(self, snapshot):
        logpsi = self.adaptor.call_network(self.params, self.data, self.system)
        assert logpsi == snapshot

    def test_potential(self, snapshot):
        potential = self.adaptor.call_local_potential_energy(
            self.params, None, self.data, self.system
        )
        assert potential == snapshot

    def test_kinetic(self, snapshot):
        kinetic = self.adaptor.call_local_kinetic_energy(
            self.params, None, self.data, self.system
        )
        assert kinetic == snapshot

    def test_walk(self, snapshot):
        # This one is pmapped
        params, data, system, aux_data = self.adaptor.restore(self.net_retore)
        key = jax.random.PRNGKey(42)
        key = jax.random.split(key, 1)
        mcmc_step = self.adaptor.make_walking_step(1, system)
        data = mcmc_step(key, params, data, aux_data)[0]
        assert jnp.array(data) == snapshot  # ShardedDeviceArray -> DeviceArray
