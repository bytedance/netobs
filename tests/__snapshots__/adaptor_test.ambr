# serializer version: 1
# name: TestFermiNet.test_kinetic
  Array(0.2784566, dtype=float32)
# ---
# name: TestFermiNet.test_potential
  Array(-0.7828356, dtype=float32)
# ---
# name: TestFermiNet.test_walk
  Array([[[0.4166998, 0.2729937, 0.9946592],
          [-0.1636449, 0.5025202, -0.2179334],
          [0.5774042, 0.5321482, -0.83614],
          [-0.6533463, 0.2915913, 1.5937456],
          [0.479838, -0.0993455, -0.2716153],
          [-0.4078843, -2.4394894, 0.041675],
          [0.0554786, 0.2439971, -0.0937411],
          [0.2253472, 0.4246373, -0.7441298],
          [0.3037185, 1.1247488, -0.7110634],
          [0.0041618, -0.7565254, -1.5015912]]], dtype=float32)
# ---
# name: TestFermiNet.test_wf
  Array(0.1269548, dtype=float32)
# ---
# name: TestHydrogen.test_kinetic
  Array(1.0260653, dtype=float32)
# ---
# name: TestHydrogen.test_potential
  Array(-1.5260653, dtype=float32)
# ---
# name: TestHydrogen.test_walk
  Array([[[-0.0269817, -0.0194543, 0.308942],
          [0.4157366, -1.0177534, 0.7061033],
          [1.1319326, -1.5159024, -1.4163642],
          [-0.5633768, 0.0473477, 1.8883846],
          [0.062221, -0.3534039, -0.3654794]]], dtype=float32)
# ---
# name: TestHydrogen.test_wf
  Array(-0.65528, dtype=float32)
# ---
# name: TestDeepSolid.test_kinetic
  DeviceArray(0.73921335-0.j, dtype=complex64)
# ---
# name: TestDeepSolid.test_potential
  DeviceArray(-2.1559424, dtype=float32)
# ---
# name: TestDeepSolid.test_walk
  DeviceArray([[[ 2.2995605 ,  0.42593208,  0.27654204,  3.4519253 ,  1.6989225 ,  0.1703778 ],
                [ 2.4035108 , 99.063805  , 99.72556   ,  0.3563191 , 99.54929   , 99.8651    ],
                [ 1.7425002 ,  1.3832064 , 99.18426   ,  3.1609056 , 99.914474  ,  0.52654046],
                [ 1.1910352 , 99.932     , 99.03048   ,  3.0361352 , 98.8087    ,  0.84947747],
                [ 2.7835414 , 98.74707   ,  0.670912  ,  2.3160307 ,  0.9291579 ,  0.03203664],
                [ 2.1043825 , 98.84852   ,  1.2825438 ,  0.42136675, 99.75692   ,  0.06060147],
                [ 1.4776574 ,  0.3267837 ,  0.260595  ,  2.0732405 , 99.6132    ,  0.40311384],
                [ 2.4202895 ,  0.8260073 , 99.986984  ,  3.2489176 , 99.280174  ,  0.8491479 ],
                [ 2.999802  ,  0.10107677, 98.864075  ,  2.4360607 ,  0.06492138,  0.55907583],
                [ 0.05181932,  0.03363067, 99.75906   ,  0.74592364, 99.53868   , 99.66028   ]]],            dtype=float32)
# ---
# name: TestDeepSolid.test_wf
  DeviceArray(-2.0724864, dtype=float32)
# ---
